classes/user_extractor.php

```diff
@@ -76,7 +76,23 @@ public static function get_user(string $fieldname, string $fieldvalue, bool $ins
             // Check if requested field exists, required for Totara compatibility.
             $fields = array_merge(\core_user::AUTHSYNCFIELDS, ['id', 'username']);
             if (in_array($fieldname, $fields)) {
-                $user = get_complete_user_data($fieldname, $fieldvalue);
+
+                $fieldsql = " AND ". $DB->sql_equal($fieldname, ':fieldvalue', !$insensitive);
```

Comment:
There is still a lot of duplicated code here. As now we use SQL for both paths of if statement we can move actual SQL execution and user search out of it. In if else we would build SQL's parts as they are different depending on scenario. Hope that makes sense.

Reply:
Hi @dmitriim I've refactored as you were suggesting. Can you have a look and hopefully merge?

Reply:
Sorry for being a pain, but it's still have some code duplication. We basically rewrite the same SQL in both parts of if statement.

Reply:
How about something like:

`    /**
     * Get extracted from DB user.
     *
     * @param string $fieldname Field name to search by.
     * @param string $fieldvalue Field value to search by.
     * @param bool $insensitive Whether to use case insensitive match.
     *
     * @return mixed False, or A {@link $USER} object.
     */
    public static function get_user(string $fieldname, string $fieldvalue, bool $insensitive = false) {
        global $DB, $CFG;

        $user = false;
        $joins = '';
        $params['fieldvalue'] = $fieldvalue;
        $params['mnethostid'] = $CFG->mnet_localhost_id;

        if (user_fields::is_custom_profile_field($fieldname)) {

            $fieldname = user_fields::get_field_short_name($fieldname);

            $joins .= " LEFT JOIN {user_info_field} f ON f.shortname = :fieldname ";
            $joins .= " LEFT JOIN {user_info_data} d ON d.fieldid = f.id AND d.userid = u.id ";

            $fieldsql = " AND ". $DB->sql_equal('d.data', ':fieldvalue', !$insensitive);
            $params['fieldname'] = $fieldname;

        } else {
            // Check if requested field exists, required for Totara compatibility.
            $fields = array_merge(\core_user::AUTHSYNCFIELDS, ['id', 'username']);
            if (in_array($fieldname, $fields)) {
                $fieldsql = " AND ". $DB->sql_equal('u.' . $fieldname, ':fieldvalue', !$insensitive);
                $params['fieldname'] = $fieldname;

            }
        }

        if (!empty($fieldsql)) {
            $sql = "SELECT u.id
                          FROM {user} u $joins
                         WHERE u.deleted <> 1 AND
                               u.mnethostid = :mnethostid $fieldsql";

            if ($records = $DB->get_records_sql($sql, $params)) {
                if (count($records) == 1) {
                    $record = reset($records);
                    $user = get_complete_user_data('id', $record->id);
                }
            }
        }

        return $user;
    }`